src/viam/sdk/common/proto_t.hpp

```diff
@@ -0,0 +1,293 @@
+    friend bool operator==(const ProtoT& lhs, const ProtoT& rhs) {
+        return lhs.vtable_.equal_to(lhs.self_.get(), rhs.self_.get(), rhs.vtable_);
+    }
+
+    void swap(ProtoT& other) {
+        self_.swap(vtable_, other.self_, other.vtable_);
+        std::swap(vtable_, other.vtable_);
+    }
+
+    // Construct from proto value
+    // This method is trivially templated to insulate Value from our API/ABI.
+    // In a translation unit which includes <google/protobuf/struct.pb.h> you can call
+    // this function on a Value instance without specifying a template parameter and it will "just
+    // work"
+    template <typename Value = google::protobuf::Value>
+    static ProtoT from_proto_value(const Value& v) {
+        return ProtoT(&v);
+    }
+
+    // Convert to protobuf Value.
+    friend void to_proto_value(const ProtoT& t, google::protobuf::Value* v);
+
+    // Obtain integer constant representing the stored data type.
+    int kind() const {
+        return vtable_.kind();
+    }
+
+    // Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoT&);
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoT&);
+
+   private:
+    struct vtable_t {
+        void (*dtor)(void*) noexcept;
+        void (*copy)(void const*, void*);
+        void (*move)(void*, void*) noexcept;
+        void (*to_proto_value)(void const*, google::protobuf::Value*);
+        int (*kind)() noexcept;
+        bool (*equal_to)(void const*, void const*, const vtable_t&);
+    };
+
+    template <typename T>
+    struct model_t {
+        model_t(T t) : data(std::move(t)) {}
+
+        static void dtor(void* self) noexcept {
+            static_cast<model_t*>(self)->~model_t();
+        }
+
+        static void copy(void const* self, void* dest) {
+            new (dest) model_t(*static_cast<model_t const*>(self));
+        }
+
+        static void move(void* self, void* dest) noexcept {
+            new (dest) model_t(std::move(*static_cast<model_t*>(self)));
+        }
+
+        static void to_proto_value(void const* self, google::protobuf::Value* v);
+
+        static int kind() noexcept;
+
+        static bool equal_to(void const* self, void const* other, const vtable_t& other_vtable) {
+            if (model_t::kind() != other_vtable.kind()) {
+                return false;
+            }
+
+            return *static_cast<T const*>(self) == *static_cast<T const*>(other);
+        }
+
+        static constexpr vtable_t vtable{dtor, copy, move, to_proto_value, kind, equal_to};
+        T data;
+    };
+
+    struct storage_t {
+        static constexpr std::size_t small_size =
+            sizeof(std::unordered_map<std::string, std::string>);
```

Comment:
A better name for it might be `local_storage_size`, it serves as the buffer size for the storage class. It's the size of a string-string map because the `unordered_map<string, ProtoT>` class is the largest class in our universe of types, but `ProtoT` is an incomplete type at time of declaration of the `storage` class, and `sizeof(unordered_map<K, V>)` doesn't depend on `K` or `V`. That said this could do with a comment and probably a `static_assert` to make sure what I've said holds since I think it's not guaranteed by the standard.